File: custom_reports/custom_reports/report/partywise_itemwise/partywise_itemwise.js
// Copyright (c) 2016, vals and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Partywise Itemwise"] = {
	"filters": [
		
			{
			"fieldname":"customer",
			"label":("Customer"),
			"fieldtype":"Link",
			"options":"Customer",
			},
			{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -12),
			"reqd": 1,
			"width": "60px"
			},
			{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1,
			"width": "60px"
			}



	]
};


// frappe.query_reports["Partywise Itemwise"] = {
// 	"filters": [
// 		{
// 			"fieldname":"month",
// 			"label": __("Month"),
// 			"fieldtype": "Select",
// 			"options": "Jan\nFeb\nMar\nApr\nMay\nJun\nJul\nAug\nSep\nOct\nNov\nDec",
// 			"default": ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov",
// 				"Dec"][frappe.datetime.str_to_obj(frappe.datetime.get_today()).getMonth()],
// 		},
// 		{
// 			"fieldname":"year",
// 			"label": __("Year"),
// 			"fieldtype": "Select",
// 			"reqd": 1
// 		},
// 		{
// 			"fieldname":"employee",
// 			"label": __("Employee"),
// 			"fieldtype": "Link",
// 			"options": "Employee"
// 		},
// 		{
// 			"fieldname":"company",
// 			"label": __("Company"),
// 			"fieldtype": "Link",
// 			"options": "Company",
// 			"default": frappe.defaults.get_user_default("Company"),
// 			"reqd": 1
// 		}
// 	],

// 	"onload": function() {
// 		return  frappe.call({
// 			method: "erpnext.hr.report.monthly_attendance_sheet.monthly_attendance_sheet.get_attendance_years",
// 			callback: function(r) {
// 				var year_filter = frappe.query_report_filters_by_name.year;
// 				year_filter.df.options = r.message;
// 				year_filter.df.default = r.message.split("\n")[0];
// 				year_filter.refresh();
// 				year_filter.set_input(year_filter.df.default);
// 			}
// 		});
// 	}
// }

File: custom_reports/custom_reports/report/partywise_itemwise/partywise_itemwise.py
# Copyright (c) 2013, vals and contributors
# For license information, please see license.txt

# import frappe

from __future__ import unicode_literals
import frappe

def execute(filters=None):
	conditions, filters = get_conditions(filters)
	columns = get_columns(filters)
	# columns2 = get_columns2(filters)
	# test = get_mapped_pri_records(conditions,filters)
	data = get_data(conditions,filters)

	return columns,data


def get_data(conditions,filters):
	
	
        
		
		item = frappe.db.sql(""" 

		select DISTINCT  si.customer,
				sii.item_code,
				sii.item_name,
				sii.packaging,
                SUM(DISTINCT sii.qty),
				
				sii.is_free_item,
               	SUM(DISTINCT sii.amount)  / SUM(DISTINCT sii.qty),
                SUM(DISTINCT sii.amount)



			from 
			`tabSales Invoice`si,
			`tabSales Invoice Item`sii



			where sii.parent = si.name and customer = %(customer)s
	GROUP BY sii.item_name;

 """.format(conditions=conditions), filters, as_dict=1)
		
 
		return item

def get_conditions(filters):
        conditions = ""
        

        if filters.get("customer"):conditions += " and si.customer in %(customer)s"

		
        
    
 
        return conditions,filters


 
def get_columns(filters):

	return  [
		# {
		# 	"fieldname":"customer",
		# 	"label":("Customer"),
		# 	"fieldtype":"Link",
		# 	"options":"Customer",
		# },
		{
			"label": ("Item code"),
			"fieldname": "item_code",
			"fieldtype": "Link",
			"options": "Item",
			"width": 100
		},
		{
			"label": ("Item Name"),
			"fieldname": "item_name",
			"width": 120
		},
		{
			"label": ("Pack"),
			"fieldname": "packaging",
			"width": 120
		},
		{
			"label": ("Qty"),
			"fieldname": "SUM(DISTINCT sii.qty)",
			"width": 50
		},
		{
			"label": ("Free"),
			"fieldname": "is_free_item",
			"width": 80
		},
		{
			"label": ("Avg.Rate"),
			"fieldname": "SUM(DISTINCT sii.amount)  / SUM(DISTINCT sii.qty)",
			"width": 120
		},
		{
			"label": ("Amount"),
			"fieldname": "SUM(DISTINCT sii.amount)",
			"width": 120
		}


	]
